central/complianceoperator/v2/report/manager/complianceReportgenerator/report_gen_impl.go

```diff
@@ -96,15 +101,33 @@ type ResultEmail struct {
 	Clusters   int
 }
 
-func (rg *complianceReportGeneratorImpl) ProcessReportRequest(req *ComplianceReportRequest) {
+func (rg *complianceReportGeneratorImpl) ProcessReportRequest(req *ComplianceReportRequest) error {
 
 	log.Infof("Processing report request %s", req)
+
+	snapshot, found, err := rg.snapshotDS.GetSnapshot(req.Ctx, req.SnapshotID)
+	if err != nil {
+		return errors.Wrap(err, "unable to retrieve the snapshot from the store")
+	}
+	if !found {
+		return errors.New("unable to find snapshot in the store")
+	}
+
 	data := rg.getDataforReport(req)
 
 	zipData, err := format(data.ResultCSVs)
 	if err != nil {
-		log.Errorf("Error zipping compliance reports for scan config %s:%s", req.ScanConfigName, err)
-		return
+		snapshot.GetReportStatus().RunState = storage.ComplianceOperatorReportStatus_FAILURE
+		snapshot.GetReportStatus().ErrorMsg = reportUtils.ErrReportGeneration.Error()
+		snapshot.GetReportStatus().CompletedAt = protocompat.TimestampNow()
+		if dbErr := rg.snapshotDS.UpsertSnapshot(req.Ctx, snapshot); dbErr != nil {
+			return errors.Wrap(dbErr, "unable to update the snapshot on report generation failure")
+		}
+		return errors.Wrapf(err, "unable to zip the compliance reports for scan config %s", req.ScanConfigName)
```

Comment:
I think, there are 6 places in the code where we do exactly this sequence:
- set state to failure
- set error
- set timestamp
- do upsert
- handle any errors

maybe we could pack it to a separate function and make the code a bit more compact?

Reply:
I've created a new function in `utils` for this.

Reply:
I am not te biggest fan of the word `utils`, but that is fine here ;)